### ISSUE TITLE
Create a way to change Serializing based upon the Accepts Header

### ISSUE THREAD

**Issue by jtyost2** [loadsys/CakePHP-Serializers#107]

**Comment by brianally** [comment]
On second thought, i'm not so sure about using ACCEPTS header as this is meant for content type. But a similar approach would be to use something like "X-Serialize: whatever". In Ember's application adapter:

```
import DS from "ember-data";

export default DS.ActiveModelAdapter.extend({
    namespace: "",
    headers: {
        "X-Serialize": "true"
    }
});
```

DesrializerFilter:

```
public function beforeDispatch(CakeEvent $event) {
    // get the request data
    $request = $event->data['request'];

    if ($request->header('X-Serialize')) {
        $data = $request->input('json_decode', true);

        // if the data is empty, set to an empty array
        if (empty($data)) {
            $data = array();
        }

        $deserializedData = array();
        foreach ($data as $key => $dataForKey) {
            $deserializedData = $this->deserializePostData($key, $data);
        }

        // merge the arrays of the deserialized data and request data
        $request->data = Hash::merge($request->data, $deserializedData);
    }
}
```

This seems to work for me, although i've yet to figure out how to stop the response back to the client from being mangled.

**Comment by justinyost** [comment]
I was actually planning on using the `Content-Type` Header to control de-serialization.

**Comment by justinyost** [comment]
Issue #123 was created to clarify the difference in changing serializing and de-serializing because of this and I must have forgotten to update this Issue.

PR #122 is my goal to have both in there as part of that PR.

**Comment by brianally** [comment]
What were you thinking to use for Content-type?

Further to tracking down how to stop serialization of the response, i've got the folowing in the controller:

```
$this->log($data);
$this->response->body(json_encode($data));
```

And i've added the following to the top of EmberDataSerializerView::toJSON()

```
CakeLog::write('error', __METHOD__);
CakeLog::write('error', print_r($data, true)); 
```

By the time it gets here $data is an empty array. :-(

**Comment by justinyost** [comment]
For Content Type to control Deserialization of the request I was planning on doing this:

Right here: https://github.com/loadsys/CakePHP-Serializers/blob/f/request-based-serialization/Routing/Filter/DeserializerFilter.php#L29

``` php
if (
 $request->header('Content-Type') === "application/json"
|| $request->header('Content-Type') === "application/vnd.api+json"
// any other content types that should be deserialized
) {
//normal deserialize method
} else {
// do nothing, let the request keep doing what it's doing
}
```

PR #122 will actually today control serialization of the response based upon the `Accepts` Header - it's not documented yet and obviously I wouldn't recommend using it.

You can today, control serialization using the `$renderAs` property to disable serializing, Read here for more information: https://github.com/loadsys/CakePHP-Serializers/blob/master/Docs/SERIALIZE_README.md#basic-controller-setup---serializing

This will go away post PR #122 being completed.

Any thoughts/suggestions/questions?

**Comment by justinyost** [comment]
@brianally If you want to control Serialization you can also use a different ViewClass for the Controller or Action in question and just not use the `Serializer` View Classes. View class isn't used for the request and nothing is serialized.

A project that I work on consumes this plugin, actually does that for some things in which we want to render some simple JSON responses that aren't tied to a model.

``` php
public function return_stuff() {
    $this->viewClass = 'Json';
    $data = array("stuff" => 'stuff2');
    $this->set(compact('data'));
    $this->set('_serialize', 'data');
}
```

**Comment by beporter** [comment]
@jtyost2: Might @brianally be having a problem due to the change from set(data) to set(modelName)?

I'm on my phone, I can't go into great detail here, but I wanted to mention it because I see the possibility of that being related from the contents of the posts today.

**Comment by brianally** [comment]
As with Accepts header, this wouldn't work for me because you're basing it on content-type. In my case, i want application/json as a response, but not serialised. (And request not deserialised.) I'm making ajax requests that must have a json response, but it's not used with Ember-Data. Using Accepts seemed like a good idea at first until i realised that the content-type is the same. This is why i was suggesting X-Serialise, or something similar. It's not sexy but it seems a bit easier to manage as there's just one place to set it up in the Ember adapter.

**Comment by justinyost** [comment]
I think I'm confused so I'm asking some questions @brianally.

You have a request, with HTTP Headers of `Content-Type: application/json` and `Accepts: application/json`

You want CakePHP-Serializers to be able to skip deserializing the request body and skip serializing the response body.

Am I right?

**Comment by brianally** [comment]
Yes, that's correct. For example, i'm making some ajax requests for some OAuth stuff. Also, a typeahead widget. Neither use ED.

In AppController's $components array i have:

```
'RequestHandler' =>array(
    'viewClassMap' => array(
        'json' => 'Serializers.EmberDataSerializer'
    )
)
```

I haven't figured out how to override that in my controller actions to keep the response from being serialised.

I have been able to skip deserialization of the request in beforeDispatch() by wrapping everything in:

```
$request = $event->data['request'];
if ($request->header('X-Serialize')) {
...
}
```

**Comment by brianally** [comment]
Here's an example where i've tried to override RequestHandler's viewClassMap so that the plugin's view is not invoked.

```
public function beforFilter() {
    parent::beforeFilter();
    $this->layout = false;
    $this->autoRender = false;
    $this->response->type('json');
}

public function autocomplete() {
    $q = $this->request->query['q'];
    $output = array();

    if (!empty($q)) {
        $output = $this->Agency->autocomplete($q);
    }
    $data = json_encode($output);
    // show config
    $this->log($this->RequestHandler->viewClassMap());
    // change and show new config
    $this->log($this->RequestHandler->viewClassMap('json', 'Json'));

    $this->response->statusCode(200);
    $this->response->body($data);
}
```

the log:

```
2015-05-21 22:30:42 Error: [{"id":"992","value":"foo"},{"id":"993","value":"foobar"}]
2015-05-21 22:30:42 Error: Array
(
    [json] => Serializers.EmberDataSerializer
    [xml] => Xml
)

2015-05-21 22:30:42 Error: Array
(
    [json] => Json
    [xml] => Xml
)

2015-05-21 22:30:42 Error: EmberDataSerializerView::toJSON
2015-05-21 22:30:42 Error: Array
(
)

2015-05-21 22:30:42 Warning: Warning (2): strlen() expects parameter 1 to be string, array given in [/.../lib/Cake/Network/CakeResponse.php, line 508]
```

I'm logging $data at the top of toJSON().

If overriding viewClassMap worked then i guess that could suffice. But, if we're going to use a header to decide whether to deserialize (or not -- the header could easily be set to stop the action) then we may as well do the same to stop serialization. At this point i'm just trying to figure out where, in the plugin code, that header check could be added. I'm mystified as to why resetting viewClassMap does not work.

**Comment by brianally** [comment]
OK, this, along with a regualr JSON view, finally bypasses EmberDataSerializerView:

```
$this->RequestHandler->viewClassMap('json', 'Json');
$this->RequestHandler->renderAs($this, 'json');
$this->set('data', $output);
$this->set('_serialize', 'data');
$this->response->statusCode(200);   
```

That works for now. I'll continue tryingto figure out a way to bypass within the plugin itself.'

I've been using Cake for ~10 years; i'm not sure where i got it in my head that i could simply set the response body and not have to use a JSON view.

**Comment by justinyost** [comment]
As an example we do this as well for passing back the OAuth tokens and such and follow this pattern:

```
public function return_stuff() {
    $this->viewClass = 'Json';
    $data = array("stuff" => 'stuff2');
    $this->set(compact('data'));
    $this->set('_serialize', 'data');
}
```

This doesn't disable Deserialization of a request.

Regarding a solution to this problem as part of CakePHP Serializers:

Serializing the response body: I think the solution of simply using a different View Class for views or controllers that you want to return JSON (or whatever) but not use CakePHP-Serializers is going to be my suggested solution.

Deserializing the request body: This is a reasonable request and I'll give it some thought and think about a good solution to this problem.
